src/hotspot/share/opto/regmask.hpp

```diff
@@ -264,115 +583,315 @@ class RegMask {
+    unsigned int lwm = MAX2(_lwm, rm._lwm);
+    for (unsigned int i = lwm; i <= hwm; i++) {
+      rm_word(i) &= ~rm.rm_word(i);
+    }
+    // If rm is smaller than our high watermark and has the infinite_stack flag set,
+    // we need to set all bits in the gap to 0.
+    if (rm.is_infinite_stack() && _hwm > rm.rm_word_max_index()) {
+      set_range(rm.rm_size_in_words(), 0, _hwm - rm.rm_word_max_index());
+      _hwm = rm.rm_word_max_index();
+    }
+    set_infinite_stack(is_infinite_stack() && !rm.is_infinite_stack());
+    trim_watermarks();
+    assert(valid_watermarks(), "sanity");
+  }
+
+  // Subtract 'rm' from 'this', but ignore everything in 'rm' that does not
+  // overlap with us and do not modify our infinite_stack flag. Supports masks of
+  // differing offsets. Does not support 'rm' with the infinite_stack flag set.
+  void SUBTRACT_inner(const RegMask& rm) {
+    assert(valid_watermarks() && rm.valid_watermarks(), "sanity");
+    assert(!rm.is_infinite_stack(), "not supported");
+    // Various translations due to differing offsets
+    int rm_index_diff = _offset - rm._offset;
+    int rm_hwm_tr = (int)rm._hwm - rm_index_diff;
+    int rm_lwm_tr = (int)rm._lwm - rm_index_diff;
+    int rm_rm_max_tr = (int)rm.rm_word_max_index() - rm_index_diff;
+    int rm_rm_size_tr = (int)rm._rm_size_in_words - rm_index_diff;
+    int hwm = MIN2((int)_hwm, rm_hwm_tr);
+    int lwm = MAX2((int)_lwm, rm_lwm_tr);
+    for (int i = lwm; i <= hwm; i++) {
+      assert(i + rm_index_diff < (int)rm._rm_size_in_words, "sanity");
+      assert(i + rm_index_diff >= 0, "sanity");
+      rm_word(i) &= ~rm.rm_word(i + rm_index_diff);
     }
+    trim_watermarks();
+    assert(valid_watermarks(), "sanity");
+  }
+
+  // Roll over the register mask. The main use is to expose a new set of stack
+  // slots for the register allocator. Return if the rollover succeeded or not.
+  bool rollover() {
+    assert(is_infinite_stack(), "rolling over non-empty mask");
+    if (!OptoRegPair::can_fit((_rm_size_in_words + _offset + _rm_size_in_words) * BitsPerWord - 1)) {
+      // Ensure that register masks cannot roll over beyond the point at which
+      // OptoRegPair can no longer index the whole mask.
+      return false;
+    }
+    _offset += _rm_size_in_words;
+    Set_All_From_Offset();
+    return true;
   }
 
   // Compute size of register mask: number of bits
-  uint Size() const;
+  uint Size() const {
+    uint sum = 0;
+    assert(valid_watermarks(), "sanity");
+    for (unsigned i = _lwm; i <= _hwm; i++) {
+      sum += population_count(rm_word(i));
+    }
+    return sum;
+  }
 
 #ifndef PRODUCT
+private:
+  bool dump_end_run(outputStream* st, OptoReg::Name start,
+                    OptoReg::Name last) const;
+
+public:
+
+  // ----------------------------------------------------------------------
+  // The methods below are only for testing purposes (see test_regmask.cpp)
+  // ----------------------------------------------------------------------
```

Comment:
I wonder if it could be solved with `friend` instead, so it does not have to be public and get accidentally used somehow.

Or maybe some `gtest_` prefix? Not sure.

Reply:
I like adding a `gtest_` prefix, I'll do that. Not sure how to make gtests work with `friend`.